src/outputs/fifo.c

```diff
@@ -0,0 +1,498 @@
+/*
+ * Copyright (C) 2016 Christian Meffert <christian.meffert@googlemail.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ */
+
+
+#ifdef HAVE_CONFIG_H
+# include <config.h>
+#endif
+
+#include <errno.h>
+#include <fcntl.h>
+#include <inttypes.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+#include <event2/event.h>
+
+#include "conffile.h"
+#include "logger.h"
+#include "player.h"
+#include "outputs.h"
+
+#define FIFO_BUFFER_SIZE 65536 /* pipe capacity on Linux >= 2.6.11 */
+
+struct fifo_session
+{
+  enum output_device_state state;
+
+  char *path;
+
+  int input_fd;
+  int output_fd;
+
+  int created;
+
+  struct event *deferredev;
+  output_status_cb defer_cb;
+
+  /* Do not dereference - only passed to the status cb */
+  struct output_device *device;
+  struct output_session *output_session;
+  output_status_cb status_cb;
+};
+
+/* From player.c */
+extern struct event_base *evbase_player;
+
+static struct fifo_session *sessions;
+
+/* Forwards */
+static void
+defer_cb(int fd, short what, void *arg);
+
+
+/* ---------------------------- FIFO HANDLING ---------------------------- */
+
+static void
+fifo_delete(struct fifo_session *fifo_session)
+{
+  DPRINTF(E_DBG, L_FIFO, "Removing FIFO \"%s\"\n", fifo_session->path);
+
+  if (unlink(fifo_session->path) < 0)
+    {
+      DPRINTF(E_WARN, L_FIFO, "Could not remove FIFO \"%s\": %d\n", fifo_session->path, errno);
+      return;
+    }
+
+  fifo_session->created = 0;
+}
+
+static void
+fifo_close(struct fifo_session *fifo_session)
+{
+  struct stat st;
+
+  if (fifo_session->input_fd >= 0)
+    {
+      close(fifo_session->input_fd);
+      fifo_session->input_fd = -1;
+    }
+
+  if (fifo_session->output_fd >= 0)
+    {
+      close(fifo_session->output_fd);
+      fifo_session->output_fd = -1;
+    }
+
+  if (fifo_session->created && (stat(fifo_session->path, &st) == 0))
+    fifo_delete(fifo_session);
+}
+
+static int
+fifo_make(struct fifo_session *fifo_session)
+{
+  if (mkfifo(fifo_session->path, 0666) < 0)
+    {
+      DPRINTF(E_LOG, L_FIFO, "Could not create FIFO \"%s\": %d\n", fifo_session->path, errno);
+      return -1;
+    }
+
```

Comment:
Consider a DBG log message here telling that the FIFO was created (like you have for removing)